### ISSUE TITLE
Cleaning up motor `delay` values

### ISSUE THREAD

**Issue by broofa** [JohnCoker/thrustcurve3#33]
FYI, I found myself [writing some code](https://github.com/broofa/thrustcurve-db/blob/main/util.js#L10-L61) to try to clean up the `delays` values in the DB.  I don't how much interest this is to you, but I thought you might want to apply some of this to the DB data, like the whitespace removal, the enforcement of "," (instead of "-") for lists, and the code that collapses sequential values to ranges (e.g. "3,4,5,6,7" becomes "3-7").

The S, M, and L conversion to ranges is probably of lessor interest.

FWIW, I'm not entirely sure if I'll keep this code in the `thrustcurve-db` module.  There's probably some value in not getting too fancy, and just having the the JSON mirror exactly what's on the website.  It's easy enough to bundle this `normalizeDelays` function in my client code.

FWIW, here are the actual changes this makes when run across the DB:

| Before | After|Comment|
|---|---|---|
| 0, 4, 6, 8 | 0,4,6,8 |
| 16, 13, 11, 9, 7 | 7,9,11,13,16 | Because why not descending order! |
| 2,3,4,5,6,7 | 2-7 |
| 2,3,4,5,6,7,8,9,10 | 2-10 |
| 23/P | 23,P | No idea what to make of this |
| 3,4,5,6,7,8 | 3-8 |
| 3,4,5,6,7,8,9,10 | 3-10 |
| 3,4,5,6,7,8,9,10,11,12 | 3-12 |
| 3,4,5,6,7,8,9,10,11,12,13 | 3-13 |
| 4, 7, 10 | 4,7,10 |
| 4,5,6,7,8,9,10,11 | 4-11 |
| 4,5,6,7,8,9,10,11,12,13 | 4-13 |
| 4-6-8 | 4,6,8 |
| 4-6-9 | 4,6,9 |
| 5,6,7,8,9,,10,11,12,13 | 5-13 |
| 5,6,7,8,9,10,11,12,13 | 5-13 |
| 5,6,7,8,9,10,11,12,13,14 | 5-14 |
| 5,6,7,8,9,10,11,12,13,14, 15, 16,17 | 5-17 |
| 5,6,7,8,9,10,11,12,13,14,15 | 5-15 |
| 5,6,7,8,9,10,11,12,13,14,15,16,17,18 | 5-18 |
| 5,6,7,8,9,11,12,13,14 | 5-9,11-14 | Missing "10"??? |
| 5-7-11-14 | 5,7,11,14 |
| 5-9-13 | 5,9,13 |
| 6,7,8,9,10,11,12,13,14 | 6-14 | Seems odd a range this wide would start at "6".  Maybe RMS drillable delay that should just be "S, M, L" |
| 6,7,8,9,10,11,12,13,14,15 | 6-15 |
| 6,7,8,9,10,11,12,13,14,15,16 | 6-16 |
| 6-10-14 | 6,10,14 | Probably RMS, so should be "S, M, L" |
| 6-8-10-12-14 | 6,8,10,12,14 |
| 7,10,13,0-14 | 0-14 | Specific delays are redundant with "0-14" |
| 7,8,9,10,11,12,13,14,15 | 7-15 |
| 7,8,9,10,11,12,13,14,15,16 | 7-16 |
| 7,8,9,10,11,12,13,14,15,16,17 | 7-17 |
| 8,9,10,11,12,13,14,15,16,17,18 | 8-18 |
| 9,10,11,12,13,14,15,16,17,18,19 | 9-19 |
| L | 0-14 |
| M | 0-10 |
| M,L | 0-14 |
| S,L,P | 0-14,P | No "M"? |
| S,M,L | 0-14 |
| S,M,L,P | 0-14,P |

**Comment by JohnCoker** [comment]
Nice work, I'll find a way to apply it to the master DB to clean up the delay values.

Are these valid for all manufacturers?

**Comment by broofa** [comment]
FYI, I've refactored that code a bit.  I've reverted the`delays` values back to what's currently in your DB and, instead have added a couple utility functions to the `thrustcurve-db` module for un/parsing these values.  Code is now here: https://github.com/broofa/thrustcurve-db/blob/main/util.js#L10-L61

> Are these valid for all manufacturers?

I believe so.  As I said, I ran this across all of the motors in DB and posted the complete list of changes, above.

The one thing I would be leery of on your end is the S/M/L delay letters.  I scanned the database and they appear to only be used for Aerotech RMS motors, and "Kosdon by Aerotech" motors.  I'm familiar enough with RMS motors, and aerotech delays are well-documented enough to know that the delay ranges I map those to are correct in that case.  But I'm not familiar with Kosdon motors at all.  I assume (???) they share the same delay system (???), but couldn't find any documentation to that effect.

So, if you're going to clean up the DB, it may make sense to leave those as-is.

All the above said, my overall sense is that the `delays` information is not particularly accurate.  Many of these values have me scratching my head a bit. (I've added detailed comments to the list, above).  Like I should go look at what motors these are on and maybe the manufacturer specs to understand what they actually mean.

**Comment by broofa** [comment]
FYI, It looks like Cesaroni encodes the maximum delay into the `designation` field, which is much more reliably provided in the TC DB than the `delays field.  (A lot of the corrections above are for Cesaroni motors).  So I'm adding this snippet to my build script:
```javascript
    // Parse `delays` out of `designation` for Cesaroni motors, as this appears to
    // be more accurate.  (Note: Cesaroni delays are adjusted with the Pro38-DAT
    // tool that allows for removing 3, 5, 7, or 9 seconds of delay
    if (motor.manufacturerAbbrev === 'Cesaroni') {
      let delay = /-(\d+)/.test(motor.designation) && RegExp.$1;
      if (delay) motor.delays = `${delay-9},${delay-7},${delay-5},${delay-3},${delay}`
    }
```

Perhaps consider doing something similar on your end?

**Comment by JohnCoker** [comment]
I looked at the description of the CTI DAT tool and it says it only to works with the Pro38 and Pro54 motors. However, the G80 has the designation `G80-14A` so presumably it's adjustable also.

**Comment by broofa** [comment]
It looks like there's a [38->29mm adapter](https://www.apogeerockets.com/Rocket_Motors/Cesaroni_Accessories/29mm_Adapter_Ring_for_38mm_Pro-DAT) that can be used to drill 29mm motors with the 38mm DAT.

While researching the above, I realized Cesaroni sells two different DAT tools:
  * The [PRO-38 DAT](https://www.apogeerockets.com/Rocket_Motors/Cesaroni_Accessories/Pro-38_Delay_Adjustment_Tool) allows for removing 3, 5, 7, and 9 seconds of delay
  * The [PRO-54 DAT](https://www.apogeerockets.com/Rocket_Motors/Cesaroni_Accessories/Pro-54_Delay_Adjustment_Tool) allows for removing any 0-10 seconds of delay.  (I'm not 100% sure of that 10 second # for the maximum amount you can remove.  It's based on this photo, which _I think_ shows "-10" seconds on the right side of the label.  'Should probably confirm this at some point.) 
  
The upshot of this is that I need to update my delay functions to mutate the delay amounts based on case size I guess.  *sigh*.

**Comment by broofa** [comment]
... and the [38-24mm adapter](https://www.apogeerockets.com/Rocket_Motors/Cesaroni_Accessories/24mm_Adapter_Ring_for_38mm_Pro-DAT) for drilling 24mm motor delays.

I.e. 38mm and smaller motors use the PRO-38 tool (0, -3, -5, -7, and -9 adjustments).  54mm use the PRO-54 tool (0 to -10 adjustments).  All other [larger] motors are plugged, so not of interest here.

**Comment by JohnCoker** [comment]
Should we start a post on TRF to confirm the PRO-54 DAT delay adjustments?

**Comment by broofa** [comment]
I'm a little embarrassed to admit I don't really frequent the TRF forum.  No particular reason, other than I haven't had a compelling reason to dive down that rabbit hole yet.

**Comment by JohnCoker** [comment]
No problem, I started the conversation: https://www.rocketryforum.com/threads/motor-delay-adjustment.167311/

**Comment by JohnCoker** [comment]
Already getting some good replies. I'll give it another day or two and then implement the rules once they're clear.
(I agree that the data should be correct in the base site, not post-processed in the `thrustcurve-db` module.)

**Comment by broofa** [comment]
Looks good, thanks!   I've responded to some of the comments there, but it says "awaiting moderation".  Probably something to do with me not having posted there before.

**Comment by broofa** [comment]
Looks like the comments there are starting to peter out.  My takeways from that conversation...

* Our understanding of how the tools work is mostly correct.  Namely ...
  * Aerotech tools (all three of them) allow for removing 2, 4, 6, or 8 seconds of delay. (I've recently used two of the three, so I'm pretty confident here)
  * Ceseroni PRO38 DAT tool removes 3, 5, 7, or 9 seconds on 24-38mm motors
  * Cesaroni PRO54 DAT tool removes 1-10 seconds for 54mm motors
* Common wisdom is that short delays can be problematic, but there's no technical limit (Nothing in the drill tools prevents users from creating short or negative delays.)
  * Aerotech advises to not have delays < 6 seconds (§1.2 [here](http://www.aerotech-rocketry.com/uploads/76b1c22b-afa5-4281-895f-fd7507a06752_29-54mm%20DMS%20Instructions%2012-29-15%20Small.pdf))
  * I can't find any official minimum limit for Cesaroni motors.  @heada [mentions](https://www.rocketryforum.com/threads/motor-delay-adjustment.167311/page-2#post-2158712) a 2 second limit, but in the same breath says [incorrectly?] that the limit for AT motors is 4 seconds, so I'm not sure credible that is.

The minimum delay issue is an interesting one.  I'm not sure how best to incorporate that here.   I guess there are two starting points for this:
1. "**TC delays indicate what is technically possible with the official drill tools**".  This would allow for small, or even negative delays down to `-[motor burn time]` (the case where you drill a hole all the way through the delay grain).
2. "**TC delays indicate only what is possible _and advisable_ with the official drill tools**".  This would limit delays to >= 6 seconds for AT and, I guess, >= 2 seconds for CES.  (Although I'd really like to see an official source for that CES minimum #. The word-of-mouth basis we have currently is pretty flimsy.)

I'm curious what your thoughts are on this.

The two use cases I have for this information - OpenRocket and flightcard.org (the app I'm writing) -  treat motor delays as more _user-driven_ rather than _user-driving_  I.e. The user is _informing_ the app about the delay they have, rather than _consulting_ the app for what delay is allowed.

Thus, I'm kind of inclined to go with the first option.  Have TC indicate the possible delays and leave it to the apps consuming this data to how, or if, the "advisable" minimums should be enforced.  But with the caveat that negative delays a pretty obviously ludicrous and confusing.  Technically possible, but not of any real value to anyone.  So require delays be at least >= 0. 

That said... if the TC data only indicates the _advisable_ delays, this doesn't prevent downstream apps from allowing other values.  For example, in flightcard.org, I use a free-form text field but with the TC delays as a "datalist".  So the TC delays just appear as suggestions.  The user can still enter whatever value they want, though.

![image](https://user-images.githubusercontent.com/164050/126073533-098d7130-0e2b-4f6b-a339-1824fb732d4c.png)

**Comment by JohnCoker** [comment]
I don't feel comfortable showing delays below what the manufacturer recommends on ThrustCurve.org. Once we find out what CTI's recommendations are, I'll deploy code like what you posted originally to update the delays for AT and CTI.

**Comment by JohnCoker** [comment]
OK, the available delays are listed with the motors on the CTI web site. For example: http://pro38.com/products/pro24/motor.php

**Comment by broofa** [comment]
Good catch.

It looks like the PRO38-adjustable motors are _generally_ constrained to a minimum delay of 2 seconds.  However there are some oddities in their data I'm not sure what to do with...
* The 168H54-10A delays are "10,7,5,3", but the  186H42-10A delays are "10,7,5,3,1".  Why the difference?
* The 68F240-15A, 1008J420-15A, and 1115J530-15A motors all have delays of "15, 12, **9**...".  Shouldn't that be "15, 12, 10,..." (based on the dat adjustment options of "0, -3, -5...")?

The PRO54-adjustable motors appear to all be drillable down to the full -10 seconds(even the 465I150-11A can be adjusted down to 1 second delay).

While I'm here, on the Aerotech front I found [this Sirius Rocketry page](https://www.siriusrocketry.biz/ishop/18-38mm-rms-delay-drilling-tool-667.html) that mentions the minimum delay for RMS motors is 4 seconds.  So the AT minimum is either 4 or 6 seconds, depending on motor type?  Or who you want to believe?  🤷

**Comment by JohnCoker** [comment]
I was planning to write something to scrape those CTI pages to update ThrustCurve.org. So, if you find any oddities, please report them to CTI. I'm not sure the best way, though.

The 6s came from an official source, but it's actually either 4 or 6s depending on the motor: https://www.rocketryforum.com/threads/aerotech-open-thread.125657/page-144#post-2158590

**Comment by JohnCoker** [comment]
Then there is Loki, which seems to use the CTI DAT, but get different amounts of reduction on different motors and also has different minimums.
https://www.rocketryforum.com/threads/loki-cti-dat.96381/
https://www.rocketryforum.com/threads/loki-delay-with-cesaroni-dat.109071/

**Comment by broofa** [comment]
> OK, the available delays are listed with the motors on the CTI web site. For example: http://pro38.com/products/pro24/motor.php

Be aware I've found some inconsistencies between the delays listed in the summary table(s), and delays shown on [the individual motor data sheets](http://pro38.com/products/pro24/motor/MotorData.php?prodid=68F240-15A).

**Comment by JohnCoker** [comment]
I manually entered the delays for all Loki 38mm motors (taken from the instruction sheets), so at least those are covered.

**Comment by broofa** [comment]
FYI, I've put together a node script for scraping delay values off the pro38.com site: https://github.com/broofa/thrustcurve-db/blob/main/build/fetch_cesaroni.js

**CAUTION: This may break the pro38.com site for you!** pro38.com appears to be [permanently?!?] blacklisting IP's that exceed some request rate limit, so use with caution!  I'm not sure what the limit is, but it's fairly low and I've definitely triggered it.  pro38.com hasn't been accepting HTTP connections from devices on my home network for the past 24 hours.  (But accessing over a VPN works.) 🤣  

The script just writes the unaltered delay values it scrapes out to a file.  If you want to save yourself the hassle, here's [what the end result looks like](https://gist.github.com/broofa/0814dcadefa50bdd7c9566afb3333462https://gist.github.com/broofa/0814dcadefa50bdd7c9566afb3333462).    You'll need to decide what to do with those values, though.  E.g. I'm currently incorporating them into `thrustcurve-db` via [this code](https://github.com/broofa/thrustcurve-db/blob/main/build/build.js#L84-L103).

**Comment by JohnCoker** [comment]
@broofa I'd like to use the data you scraped already, but the link is either dead or I don't have permission to view it.

**Comment by broofa** [comment]
Sorry, I fat-fingered the URL.  Fixed it, above, but here it is just in case: https://gist.github.com/broofa/0814dcadefa50bdd7c9566afb3333462

**Comment by JohnCoker** [comment]
OK, I applied your algorithm and scraped data to CTI motors. This also updates a few designations to be consistent with the CTI site.

**Comment by broofa** [comment]
Looks good.  I've removed the Cesaroni logic from the thrustcurve-db build script, so it should just reflect what's in your DB now.  Published as thrustcurve-db@0.3.13.

Any interest in normalizing delay values like this:
E.g.
```
2,3,4,5,6,7 -> 2-7   # convert sequential values to ranges
7,10,13,0-14 -> 0-14  # overlapping values
16, 13, 11, 9, 7 -> 7,9,11,13,16 # out-of-order values, whitespace
```

The `parseDelays()` function in thrustcurve-db can take care of this.  But it also expands the "S", "M", and "L" aerotech values, so maybe skip this step for delays with those sorts of values?

**Comment by JohnCoker** [comment]
I think we should clean up the existing delays in general, and fix the AT delays in particular now that we know the range of adjustment.

I'm leery of ranges (`-`) though. RASP files use dashes to separate individual delay values (like we use commas), so I think it would be confusing.

**Comment by broofa** [comment]
> I think we should clean up the existing delays in general, and fix the AT delays in particular now that we know the range of adjustment.

Oh, cool.  I wasn't sure you wanted to do that.  And, yeah, the letter delays are problematic due to it being possible to drill the "M" and "S" delays to shorter-than-recommended values.

> RASP files use dashes to separate individual delay values (like we use commas), so I think it would be confusing.

It's already confusing.  The DB is using dashes as both [ranges](https://www.thrustcurve.org/motors/AeroTech/G80T/) and [value separators](https://www.thrustcurve.org/motors/AeroTech/H130W/), leading to at least [one ambiguous case](https://www.thrustcurve.org/motors/Quest/B4/).  ("Does '4-6' mean '4,5,6' or '4,6'?").

Re: RASP format - Only 7 motors in the DB use the RASP hyphen notation (9 if you count the ambiguous "4-6" case).  But there are only 29 motors with dash-as-range notation.  So neither approach is ubiquitous enough make a compelling argument for one way or the other.

Personally, I like the terseness and expressiveness of the comma-and-dash notation I'm using in thrustcurve-db.  It's a little bit more work to parse, but I've already written code for that.

**Comment by JohnCoker** [comment]
Yeah, I agree we should clean up the DB, along the lines of your original post. I believe any current dashes are just separators, not ranges, so something like your original script makes sense as a first pass.

We just expanded the CTI delays to individual values (no ranges), and I think that's cleaner and simpler. I'm not totally against the range syntax, but it seems to me to be unnecessary complexity.

**Comment by broofa** [comment]
> I'm not totally against the range syntax, but it seems to me to be unnecessary complexity.

In looking closer at the DB, I think you're right.  In practice, there just aren't that many motors that would actually have a range in `delays`.  I think it's just the Cesaroni Pro54 motors, where the Pro54-DAT allows you to adjust the delay in 1-second increments.  There's only 50 of those in the DB.

I would still prefer to see commas instead of dashes, though.

**Comment by JohnCoker** [comment]
Agreed, I'll deploy something to clean up the existing values and expand the AeroTech letter delays to numbers.

**Comment by JohnCoker** [comment]
OK, I cleaned up general delay values as per your original suggestions.

In addition, since we know no delays are adjustable down to 0s, I replaced all values like "0-14" with just "14". A subsequent pass will apply the rules for AeroTech delays, but at least all delay values now are in a consistent format:
 - spaces removed
 - only separated by commas
 - always increasing in value

**Comment by JohnCoker** [comment]
OK, AT delays have also been fixed. The letters have been turned into numbers and for 54mm and smaller motors, the delay adjustments have been added (taking into account the SU and reload minimum values).

**Comment by JohnCoker** [comment]
I believe the delay values in the database now have been standardized and include all the achievable delays where adjustable.

**Comment by broofa** [comment]
Looks good! Thanks for doing this.
Published in `thrustcurve-db@0.3.14`

![image](https://media3.giphy.com/media/SQgbkziuGrNxS/giphy.gif?cid=ecf05e47dobfiozp1kzi9k0wjyoslyamf61b805rbbadeu9a&rid=giphy.gif&ct=g)
